Feature/datastructure

funketh opened the pull request:
#14

funketh committed:
Revert "Rename Anchor to Node"

This reverts commit 16e745bda189093b141fd605a99009b2f37f3f6b.

funketh committed:
remove unnecessary constants field as CONSTANTS now has static fields

funketh committed:
update .gitignore to ignore build results and other unneeded files

funketh committed:
created base vertex class for all sections to inherit from

- added System.Collections.Immutable package
- every vertex has an ImmutableArray of its anchors
    - an exception is raised if the amount of anchors in an array is not valid for a section because specifying the amount of elements in an array is not possible in type signatures

funketh committed:
formatting

funketh committed:
add NuGet package manager plugin for Unity (https://github.com/GlitchEnzo/NuGetForUnity)

remove old packages.config (packages are now installed with NuGetForUnity)

install `System.Collections.Immutable` with NuGetForUnity

no longer ignore *.meta files

funketh committed:
move sections/ vertices into their own folder

funketh committed:
add `parentRoad` field to Anchor

Commit:
Make second constructor of `TeeSection` public

funketh committed:
rename TeeSection anchor fields (https://en.wikipedia.org/wiki/Lane#Types_of_lanes)

funketh committed:
fix `Road` and `Anchor`, lanes are now objects

- `Lane` objects can have a combination of `Lanetype`s
- lane count can be obtained from the length of the `ImmutableArray` saved in `Road`
- anchors are now identified by `AnchorNumber`s in a `Dictionary`
    - use `AnchorNumber` enum instead of `int` for static type checking
- `Anchor`s can be constructed without a parent to avoid the cyclic construction problem

Shafma committed:
Remove superfluous script `Pathfinding`

funketh committed:
move DataTypes related to roads from `DataTypes/` to `DataTypes/Edges/`

Shafma commented on Assets/Scripts/DataTypes/Edges/Anchor.cs:
```diff
@@ -0,0 +1,28 @@
+{
+    public struct Anchor
+    {
+        public Road parentRoad;
```
Straßen sind hiernach nie mit Straßen verbunden.
Diese Annahme gilt es wahrzunehmen.

funketh replied:
In #14 festgehalten: https://github.com/rrickfox/traffic-sim/issues/14#issuecomment-576052655

Shafma commented on Assets/Scripts/DataTypes/Edges/Anchor.cs:
```diff
@@ -0,0 +1,28 @@
+        public Road parentRoad;
+        public AnchorNumber number;
+        public Vector2 position;
+        public ImmutableArray<Lane> endingLanes;
```
Merges und Splits müssen durch Sections symbolisiert werden.

funketh replied:
In #14 festgehalten: https://github.com/rrickfox/traffic-sim/issues/14#issuecomment-576052655

Shafma commented on Assets/Scripts/DataTypes/Vertices/CrossSection.cs:
```diff
@@ -0,0 +1,35 @@
+        }
+
+        public CrossSection(Anchor upAnchor, Anchor rightAnchor, Anchor downAnchor, Anchor leftAnchor)
+            : base(ImmutableArray.Create(upAnchor, rightAnchor, downAnchor, leftAnchor))
```
Man könnte auch den Constructor von CrossSection selbst callen, also statt base this.

funketh replied:
76029d2b25d11d6a2e8ce509161b72ffe3495fbe & 22bc498ed7eb7018bf84fc792b922c7400f4f3b8

Shafma commented on Assets/Scripts/DataTypes/Vertices/TeeSection.cs:
```diff
@@ -0,0 +1,32 @@
+        }
+
+        public TeeSection(Anchor throughAnchor1, Anchor throughAnchor2, Anchor turnAnchor)
+            : base(ImmutableArray.Create(throughAnchor1, throughAnchor2, turnAnchor))
```
Siehe CrossSection

funketh replied:
e21061cf7bfc77ea22ff44513ce4815b38b12f52

funketh replied:
76029d2b25d11d6a2e8ce509161b72ffe3495fbe & 22bc498ed7eb7018bf84fc792b922c7400f4f3b8

Shafma commented on Assets/Scripts/DataTypes/Vertices/TeeSection.cs:
```diff
@@ -0,0 +1,32 @@
+    {
+        protected Anchor throughAnchor1;
+        protected Anchor throughAnchor2;
+        protected Anchor turnAnchor;
```
Nomenklatur fragwürdig

funketh replied:
e21061cf7bfc77ea22ff44513ce4815b38b12f52

Shafma commented on Assets/Scripts/DataTypes/Vertices/TeeSection.cs:
```diff
@@ -0,0 +1,32 @@
+            throughAnchor1 = this.anchors[0];
+            throughAnchor2 = this.anchors[1];
+            turnAnchor = this.anchors[2];
+        }
```
Einschränkung der Reihenfolge der Anchor im Enumerable notwendig

funketh replied:
22bc498ed7eb7018bf84fc792b922c7400f4f3b8

funketh replied:
Für den 1. Sprint erstmal geklärt.
Weiteres Feature hier notiert: https://github.com/rrickfox/traffic-sim/issues/14#issuecomment-576053279

funketh committed:
remove redundant constructor bodies
(https://github.com/rrickfox/traffic-sim/pull/16#discussion_r368316842)

funketh committed:
`Vertex.anchors` is now private, subclasses of it are now only constructable with explicit parameters

- this removes the need to check the array's length
- `Vertex.anchors` is only needed for internal pathfinding algorithms in `Vertex`

funketh committed:
`TeeSection` field nomenclature
(https://github.com/rrickfox/traffic-sim/pull/16#discussion_r368316955)

funketh committed:
naming convention

RichardKarlFuchs commented on Assets/Scripts/DataTypes/Vertices/CrossSection.cs:
```diff
@@ -0,0 +1,35 @@
+{
+    public class CrossSection : Vertex
+    {
+        protected Anchor upAnchor;
+        protected Anchor rightAnchor;
+        protected Anchor downAnchor;
+        protected Anchor leftAnchor;
```
Ist die Orientierung der ein-/ausgehenden Straßen relevant?

Shafma replied:
Ich bin mir nicht ganz im Klaren, was du mit Orientierung meinst.

RichardKarlFuchs commented on Assets/Scripts/DataTypes/Vertices/CrossSection.cs:
```diff
@@ -0,0 +1,35 @@
+{
+    public class CrossSection : Vertex
+    {
+        protected Anchor upAnchor;
+        protected Anchor rightAnchor;
+        protected Anchor downAnchor;
+        protected Anchor leftAnchor;
```
Ist die Orientierung der ein-/ausgehenden Straßen relevant?

funketh replied:
Nein?
Straßen umfassen Spuren in beide Richtungen.

RichardKarlFuchs commented on Assets/Scripts/DataTypes/Vertices/TeeSection.cs:
```diff
@@ -0,0 +1,32 @@
+    {
+        protected Anchor throughAnchor1;
+        protected Anchor throughAnchor2;
+        protected Anchor turnAnchor;
```
Muss es stets zwei durchgehende und eine sich aufspaltende Straße geben?

Shafma replied:
So funktionieren nunmal T-Sections.

funketh replied:
So funktionieren T-Sections, ja.
Bzw. siehe die neuen Änderungen der Nomenklatur, die dürften es ersichtlicher machen.

RichardKarlFuchs commented on Assets/Scripts/DataTypes/Vertices/Vertex.cs:
```diff
@@ -0,0 +1,15 @@
+{
+    public class Vertex
+    {
+        protected ImmutableArray<Anchor> anchors;
```
Vertices fassen somit stets eine Kreuzung zusammen, richtig?

Shafma replied:
Nein, das kann auch einfach das Ende von einer Straße sein. Ist grundsätzlich halt nur eine Ansammlung von Anchors, die u.a. Kreuzungen bilden.

funketh replied:
bis jetzt ja, aber es kann auch Unterklassen von Vertex geben, die keine wirklichen Kreuzungen sind.
z.B. Spawnpoints können Vertices sein, welche einen Anchor haben
Vorher hätten wir die halt auch Kreuzungen genannt, jetzt sind es bloß Vertices

funketh committed:
delete VertexExceptions.cs because it's no longer needed

funketh commented:
Bitte um erneutige Review des Pull Requests, damit dieser Pull Request bald gesquashed & gemerged werden kann, da alle anderen Branches von diesem abhängen.

funketh commented:
Kommentare würde ich in einem neuen Pull Request hinzufügen (hat natürlich auch hohe Priorität)

rrickfox commented:
Eröffnest du dann einen neuen Pull Request?

funketh commented:
mach ich sobald der hier durch ist

rrickfox commented:
Ich dachte jetzt an Github Kommentare...

funketh commented:
Jetzt hast du mich auch verwirrt

Shafma commented on Assets/Scripts/DataTypes/Edges/Anchor.cs:
```diff
@@ -0,0 +1,28 @@
+            : this(number, position, endingLanes)
+        {
+            this.parentRoad = parentRoad;
+        }
```
Constructor requires calling a Road-Object to assign as parentRoad, yet a Road-Object can not exist without already having two Anchors assigned directly. This would end up never getting called.

funketh replied:
8b59dd2

funketh committed:
remove unusable `Anchor` constructor
(#16 https://github.com/rrickfox/traffic-sim/pull/16#discussion_r368691265)

- this constructor is unusable because roads require their anchors to be constructed first
